### PR TITLE
Ribbon for ENS & UD screens

### PR DESCRIPTION
**Pull request by biafra23**
Handles #1538

Changes proposed in this pull request:
- show ribbon on add safe screens
- enabling chain id parameter for almost all calls to the backend
- create safe for selected network id
- **Missing** dynamic handling of UD and ENS

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt [hunk 1]
```diff
@@ -57,6 +58,8 @@ class EnsInputDialog : BaseViewBindingDialogFragment<DialogEnsInputBinding>() {
             backButton.setOnClickListener { dismiss() }
             confirmButton.setOnClickListener { onClick.offer(Unit) }
             dialogEnsInputUrl.showKeyboardForView()
+            //TODO: Replace with network name selected for safe addition (& set colors accordingly)
+            chainRibbon.text = BuildConfig.BLOCKCHAIN_NAME
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt:62]: we should have chain data here that was passed in a bundle from chain selection screen

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt [hunk 2]
```diff
@@ -58,6 +59,8 @@ class UnstoppableInputDialog : BaseViewBindingDialogFragment<DialogUnstoppableIn
             backButton.setOnClickListener { dismiss() }
             confirmButton.setOnClickListener { onClick.offer(Unit) }
             dialogUnstoppableInputDomain.showKeyboardForView()
+            //TODO: Replace with network name selected for safe addition (& set colors accordingly)
+            chainRibbon.text = BuildConfig.BLOCKCHAIN_NAME
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:63]: same here

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeFragment.kt [hunk 3]
```diff
@@ -45,12 +50,15 @@ class AddSafeFragment : BaseViewBindingFragment<FragmentAddSafeBinding>() {
...
             nextButton.setOnClickListener {
-                addSafeAddressInputLayout.address?.let { viewModel.validate(it) }
+                addSafeAddressInputLayout.address?.let { viewModel.validate(it, selectedChain) }
             }
             backButton.setOnClickListener { findNavController().navigateUp() }
             addSafeAddressInputLayout.setOnClickListener {
                 addressInputHelper.showDialog()
             }
+            chainRibbon.text = selectedChain.name
+            chainRibbon.setBackgroundColor(Color.parseColor(selectedChain.backgroundColor))
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeFragment.kt:60]: I would use try catch here (see applying colors for ribbon in StartActivity) as color parsing can fail due to possible malformed format coming from backend

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameFragment.kt [hunk 4]
```diff
@@ -48,9 +50,13 @@ class AddSafeNameFragment : BaseViewBindingFragment<FragmentAddSafeNameBinding>(
             backButton.setOnClickListener { findNavController().navigateUp() }
             nextButton.setOnClickListener {
                 addSafeNameLayout.isErrorEnabled = false
-                viewModel.submitAddressAndName(newAddress, addSafeNameEntry.text.toString())
+                viewModel.submitAddressAndName(newAddress, addSafeNameEntry.text.toString(), selectedChain)
             }
             addSafeNameEntry.doOnTextChanged { text, _, _, _ -> binding.nextButton.isEnabled = !text.isNullOrBlank() }
+
+            chainRibbon.text = selectedChain.name
+            chainRibbon.setBackgroundColor(Color.parseColor(selectedChain.backgroundColor))
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameFragment.kt:58]: same as above. parseColor could throw exception and crash the app

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameFragment.kt [hunk 5]
```diff
@@ -60,11 +66,11 @@ class AddSafeNameFragment : BaseViewBindingFragment<FragmentAddSafeNameBinding>(
                         when (action) {
                             is CloseScreen -> {
                                 requireActivity().hideSoftKeyboard()
-                                findNavController().popBackStack(R.id.addSafeFragment, true)
+                                findNavController().popBackStack(R.id.selectChainFragment, true)
                             }
                             is ImportOwner -> {
                                 requireActivity().hideSoftKeyboard()
-                                findNavController().navigate(AddSafeNameFragmentDirections.actionAddSafeNameFragmentToAddSafeOwnerFragment(binding.addSafeNameEntry.text.toString(), navArgs.newAddress))
+                                findNavController().navigate(AddSafeNameFragmentDirections.actionAddSafeNameFragmentToAddSafeOwnerFragment(binding.addSafeNameEntry.text.toString(), navArgs.newAddress, selectedChain))
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameFragment.kt:73]: owner should be chain agnostic. I think we don't need chain data there as owners are added not for a specific chain

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameViewModel.kt [hunk 6]
```diff
@@ -23,15 +24,15 @@ class AddSafeNameViewModel
...
+    fun submitAddressAndName(address: Solidity.Address, localName: String, chain: Chain) {
         safeLaunch {
             localName.takeUnless { it.isBlank() } ?: run {
                 updateState { AddSafeNameState(ViewAction.ShowError(InvalidName())) }
                 return@safeLaunch
             }
             updateState { AddSafeNameState(ViewAction.Loading(true)) }
             runCatching {
-                val safe = Safe(address, localName.trim())
+                val safe = Safe(address, localName.trim(), chain.chainId)
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeNameViewModel.kt:35]: I think we should also save chain data here. Currently it seems that chain is not saved to a db at all

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeOwnerFragment.kt [hunk 7]
```diff
@@ -70,6 +72,9 @@ class AddSafeOwnerFragment : BaseViewBindingFragment<FragmentAddSafeOwnerBinding
                 tracker.logOnboardingOwnerSkipped()
                 finishAddSafeFlow()
             }
+            chainRibbon.text = selectedChain.name
+            chainRibbon.setBackgroundColor(Color.parseColor(selectedChain.backgroundColor))
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeOwnerFragment.kt:76]: parseColor can throw

File: app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeViewModel.kt [hunk 8]
```diff
@@ -17,12 +18,12 @@ class AddSafeViewModel
 
     override fun initialState(): State = AddSafeState(ViewAction.Loading(false))
 
-    fun validate(address: Solidity.Address) {
+    fun validate(address: Solidity.Address, chain: Chain) {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/safe/add/AddSafeViewModel.kt:21]: fun validate(safe: Safe) would be more legible

File: app/src/main/java/io/gnosis/safe/ui/transactions/details/ConfirmRejectionViewModel.kt [hunk 9]
```diff
@@ -101,7 +102,7 @@ class ConfirmRejectionViewModel
         if (txDetails == null) {
             safeLaunch {
                 updateState { ConfirmationRejectedViewState(ViewAction.Loading(true)) }
-                txDetails = transactionRepository.getTransactionDetails(txId)
+                txDetails = transactionRepository.getTransactionDetails(txId, safeRepository.getActiveSafe()!!.chainId)
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/transactions/details/ConfirmRejectionViewModel.kt:105]: more natural sequence of arguments would be chainId, txId

File: app/src/main/java/io/gnosis/safe/helpers/AddressInputHelper.kt [hunk 10]
```diff
@@ -24,7 +24,8 @@ class AddressInputHelper(
     fragment: BaseFragment,
     tracker: Tracker,
     private val addressCallback: (Solidity.Address) -> Unit,
-    private val errorCallback: (Throwable, String?) -> Unit
+    private val errorCallback: (Throwable, String?) -> Unit,
+    private val selectedChain: Chain
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/helpers/AddressInputHelper.kt:28]: I would declare parameters before lambdas:
```kotlin
class AddressInputHelper(
    fragment: BaseFragment,
    tracker: Tracker,
    private val selectedChain: Chain,
    private val addressCallback: (Solidity.Address) -> Unit,
    private val errorCallback: (Throwable, String?) -> Unit
```

File: app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt [hunk 11]
```diff
@@ -22,11 +24,12 @@ import kotlinx.coroutines.flow.asFlow
 import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.launch
 import pm.gnosis.model.Solidity
+import pm.gnosis.svalinn.common.utils.getColorCompat
 import pm.gnosis.svalinn.common.utils.showKeyboardForView
 import pm.gnosis.svalinn.common.utils.visible
 import javax.inject.Inject
 
-class EnsInputDialog : BaseViewBindingDialogFragment<DialogEnsInputBinding>() {
+class EnsInputDialog(private val selectedChain: Chain) : BaseViewBindingDialogFragment<DialogEnsInputBinding>() {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt:32]: selectedChain should be passed as an argument in a bundle

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt:32]: https://developer.android.com/reference/android/app/Fragment.html#public-constructors_1

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt [hunk 12]
```diff
@@ -23,11 +25,12 @@ import kotlinx.coroutines.flow.asFlow
 import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.launch
 import pm.gnosis.model.Solidity
+import pm.gnosis.svalinn.common.utils.getColorCompat
 import pm.gnosis.svalinn.common.utils.showKeyboardForView
 import pm.gnosis.svalinn.common.utils.visible
 import javax.inject.Inject
 
-class UnstoppableInputDialog : BaseViewBindingDialogFragment<DialogUnstoppableInputBinding>() {
+class UnstoppableInputDialog(private val selectedChain: Chain) : BaseViewBindingDialogFragment<DialogUnstoppableInputBinding>() {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:33]: same here

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt [hunk 13]
```diff
@@ -58,6 +61,15 @@ class UnstoppableInputDialog : BaseViewBindingDialogFragment<DialogUnstoppableIn
             backButton.setOnClickListener { dismiss() }
             confirmButton.setOnClickListener { onClick.offer(Unit) }
             dialogUnstoppableInputDomain.showKeyboardForView()
+            chainRibbon.text = selectedChain.name
+            try {
+                chainRibbon.setTextColor(Color.parseColor(selectedChain.textColor))
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:66]: we should introduce extension function and add it to AppUtils. This code is reused in many places

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt [hunk 14]
```diff
@@ -77,38 +89,38 @@ class UnstoppableInputDialog : BaseViewBindingDialogFragment<DialogUnstoppableIn
...
-                        binding.successViews.visible(true)
-                        binding.dialogUnstoppableDomainLayout.isErrorEnabled = false
-                        onNewAddress.offer(address)
-                        addressHelper.populateAddressInfo(
-                                binding.dialogUnstoppableInputAddress,
-                                binding.dialogEnsInputAddressImage,
-                                address
-                        )
+                .onSuccess { address ->
+                    binding.dialogEnsInputProgress.visible(false)
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:93]: suggestion for improving legibility
```kotlin
.onSuccess { address ->
  with(binding) {
              dialogEnsInputProgress.visible(false)
              .....
```

File: data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt [hunk 15]
```diff
@@ -28,6 +28,8 @@ class ChainInfoRepository(
         }
     }
 
+    suspend fun update(chain: Chain) = chainDao.save(chain)
```
Comment (elgatovital) [inline data/src/main/java/io/gnosis/data/repositories/ChainInfoRepository.kt:31]: initially there will be not chain in a db (as no safes are loaded). So first call of the update will save. Maybe calling it save would better represent it's function

File: app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt [hunk 16]
```diff
@@ -149,6 +160,20 @@ class EnsInputDialog(private val selectedChain: Chain) : BaseViewBindingDialogFr
     }
 
     companion object {
-        fun create(chain: Chain) = EnsInputDialog(chain)
+        fun create(chain: Chain): EnsInputDialog {
+            val dialog = EnsInputDialog()
+            dialog.arguments = bundleOf(
+                CHAIN_NAME to chain.name,
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt:166]: you can just pass chain as a Serializable

File: app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt [hunk 17]
```diff
@@ -149,6 +160,20 @@ class EnsInputDialog(private val selectedChain: Chain) : BaseViewBindingDialogFr
...
+                CHAIN_ID to chain.chainId,
+                CHAIN_BACKGROUND_COLOR to chain.backgroundColor,
+                CHAIN_TEXT_COLOR to chain.textColor
+            )
+            return dialog
+        }
+
+        private const val CHAIN_NAME = "chain_name"
+        private const val CHAIN_ID = "chain_id"
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/EnsInputDialog.kt:175]:
```kotlin
private val ARGS_CHAIN = "args.serializable.chain"
```

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt [hunk 18]
```diff
@@ -23,12 +26,15 @@ import kotlinx.coroutines.flow.asFlow
...
 import kotlinx.coroutines.launch
 import pm.gnosis.model.Solidity
+import pm.gnosis.svalinn.common.utils.getColorCompat
 import pm.gnosis.svalinn.common.utils.showKeyboardForView
 import pm.gnosis.svalinn.common.utils.visible
 import javax.inject.Inject
 
 class UnstoppableInputDialog : BaseViewBindingDialogFragment<DialogUnstoppableInputBinding>() {
 
+    lateinit var selectedChain: Chain
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:36]:
```kotlin
private val selectedChain by lazy { requireArguments()[ARGS_CHAIN] as Chain }
```

File: app/src/main/java/io/gnosis/safe/utils/AppUtils.kt [hunk 19]
```diff
@@ -84,3 +87,13 @@ fun showConfirmDialog(
         title = if (title == null) null else context.resources.getString(title)
     ).show()
 }
+
+fun String.safeParseColorWithDefault(context: Context, defaultColor: Int, tracker: Tracker? = null): Int {
```
Comment (elgatovital) [inline app/src/main/java/io/gnosis/safe/utils/AppUtils.kt:91]: String.toColor(...  would be a better name
defaultColor -> failsafeColor
then everything should be clear and we don't need such a long name

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/utils/AppUtils.kt:91]: defaultColor should have a default value

Comment (elgatovital) [reply app/src/main/java/io/gnosis/safe/utils/AppUtils.kt:91]: also I think we don't need tracker here. Timber.e should be enough as all errors are submitted as non-fatals to firebase

File: app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt (hunk 18 above)
Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/ui/dialogs/UnstoppableInputDialog.kt:35]: fancy!

File: app/src/main/java/io/gnosis/safe/utils/AppUtils.kt (hunk 19 above)
Comment (biafra23) [reply app/src/main/java/io/gnosis/safe/utils/AppUtils.kt]: I also moved it to StringUtils

**Review by elgatovital — APPROVED** [review]

**Pull request merged**